types/braintree-web-drop-in/index.d.ts

```diff
@@ -0,0 +1,130 @@
+*/
+
+// Options
+
+export interface Options {
+    authorization: string;
+    container: any;
+    locale?: string;
+    translations?: object;
+    paymentOptionPriority?: string[];
+    card?: boolean | cardCreateOptions;
+    paypal?: paypalCreateOptions;
+    paypalCredit?: paypalCreateOptions;
+    venmo?: venmoCreateOptions | boolean;
+    applePay?: applePayCreateOptions;
+    googlePay?: googlePayCreateOptions;
+    dataCollector?: dataCollectorOptions;
+    threeDSecure?: threeDSecureOptions;
+    vaultManager?: boolean;
+    preselectVaultedPaymentMethod?: boolean;
+}
+
+export interface applePayCreateOptions {
+    buttonStyle?: string;
+    displayName: string;
+    applePaySessionVersion?: number;
+    paymentRequest: any;
+}
+
+export interface cardCreateOptions {
+    cardholderName?:
+        | boolean
+        | {
+              required?: boolean;
+          };
+    overrides: {
+        fields?: object;
+        styles?: object;
+    };
+    clearFieldsAfterTokenization?: boolean;
+    vault?: {
+        allowVaultCardOverride?: boolean;
+        vaultCard?: boolean;
+    };
+}
+
+export interface dataCollectorOptions {
+    kount?: boolean;
+    paypal?: boolean;
+}
+
+export interface googlePayCreateOptions {
+    merchantId: string;
+    googlePayVersion?: string;
+    transactionInfo: any;
+    button?: any;
+}
+
+export interface paypalCreateOptions {
+    flow: "checkout" | "vault";
+    amount?: string | number;
+    currency?: string;
+    buttonStyle?: string;
+    commit?: boolean;
+}
+
+export interface threeDSecureOptions {
+    amount: string;
+}
+
+export interface venmoCreateOptions {
+    allowNewBrowserTab?: boolean;
+}
+
+// Dropin
+
+export interface Dropin {
+    clearSelectedPaymentMethod(): void;
+    isPaymentMethodRequestable(): boolean;
+    on(event: "noPaymentMethodRequestable", handler: () => void): void;
+    on(
+        event: "paymentMethodRequestable",
+        handler: (payload: { type: "CreditCard" | "PayPalAccount"; paymentMethodIsSelected: boolean }) => void
+    ): void;
+    on(
+        event: "paymentOptionSelected",
+        handler: (payload: { paymentOption: "card" | "paypal" | "paypalCredit" }) => void
+    ): void;
+    requestPaymentMethod(callback: (error: object | null, payload: PaymentMethodPayload | undefined) => void): void;
+    requestPaymentMethod(): Promise<PaymentMethodPayload>;
+    teardown(callback: (error: object | null | undefined) => void): void;
+    teardown(): Promise<void>;
+}
+
+export interface PaymentMethodPayload {
+    nonce: string;
+    details: object;
+    type: "CreditCard" | "PayPalAccount" | "VenmoAccount" | "AndroidPayCard" | "ApplePayCard";
+    deviceData: string | null;
+    [key: string]: any;
+}
+
+// Methods
+
+export function create(options: Options, callback: (error: object | null, dropin: Dropin | undefined) => void): void;
+export function create(options: Options): Promise<Dropin>;
+
+// Global
+
+declare global {
```

Comment:
I think a more appropriate version of this would be

```ts
export as namespace braintree;
```

since it appears to be conditionally exposed as a global. Once that's fixed I think we can pull this in.

Reply:
In a module, the user would have to import ;braintree-web-drop-in', and then access it like `dropin.create()`, but globally the user would access it as an extension of Braintree's main module like so `braintree.dropin.create()`
Using `declare global` I can make the `braintree` type declaration related to the global context, not to the current module.
`export as namespace braintree` wouldn't work here since `braintree` is not the module being exported, but `dropin` is.